### PR TITLE
Persistent workspaces in hyprland/workspaces

### PR DESCRIPTION
**Pull request by zjeffer**
Adds `persistent_workspaces` to the hyprland/workspaces module, based on the implementation from wlr/workspaces.

Apart from what's possible with wlr/workspaces (specifying an array of monitors a workspace should be shown on), you can also specify a fixed number of workspaces each monitor should have, like this:

```jsonc
"persistent_workspaces": {
  "*": 3,    // 3 workspaces by default
  "DP-1": 5 // but DP-1 gets 5 workspaces
}
```

CC: @MonstrousOgre @Anakael @MightyPlaza

### CODE REVIEW

**Comment by MonstrousOgre** [comment]
This is great! Once this is merged, I can start using this over wlr/workspaces

**Comment by MightyPlaza** [comment]
I was not using window_, so I just created the var and parsed it on start.
We need to listen to windowCreate and windowDestroy (or more) to keep track of window count

**Comment by MightyPlaza** [comment]
It's possible to listen to (openwindow, movewindow, closewindow) but you'd have to keep track of which window are where for it to work.

**Comment by zjeffer** [comment]
> Seems to work pretty well, though.

Nevermind, still a small bug when moving a window to another workspace: the workspace is not seen as empty until I go back to the empty workspace.

For example, here I move a window from workspace 1 to 4, one by one:

https://github.com/Alexays/Waybar/assets/4633209/51be6e13-2073-4315-97d3-2cb04aff3f33

A white background means the workspace is not empty, grey means persistent and empty.

**Comment by zjeffer** [comment]
Ok I think I fixed it for real this time: now it will update the window counts on `openwindow`, `closewindow`, and `movewindow`, and set it to 0 when the json received from hyprland's IPC does not include the workspace (meaning it is a persistent workspace).

Ready for review.

**Comment by MonstrousOgre** [comment]
What about listening for `destroyworkspace` to know if it's empty? Don't empty workspaces get destroyed when not focused?

File: src/modules/hyprland/workspaces.cpp
Comment (zjeffer) [inline src/modules/hyprland/workspaces.cpp:180]: @MonstrousOgre If I just add `(*workspace)->set_windows(0);` here (which gets called when a persistent workspace gets a destroyworkspace event), the whole thing works without listening to the open/close/move-window events. The only problem is the windows_ variable does not show the correct amount: it always shows 0 or 1. Adding more windows does not update the window count.

I can imagine in the future, someone would want to make use of the windows_ variable for something other than knowing whether the workspace is empty or not. With this PR's current implementation, the windows_ variable shows the actual window count.

Comment (MonstrousOgre) [reply src/modules/hyprland/workspaces.cpp:180]: That makes sense

File: src/modules/hyprland/workspaces.cpp
Comment (MightyPlaza) [inline src/modules/hyprland/workspaces.cpp:228]: don't use id
this won't work for named workspaces or named special workspaces

**Review by MightyPlaza — CHANGES_REQUESTED** [review]

File: src/modules/hyprland/workspaces.cpp
Comment (MightyPlaza) [inline src/modules/hyprland/workspaces.cpp:321]: persistent workspaces makes implementing label id impossible, so this should be removed and doc updated

**Review by MightyPlaza — CHANGES_REQUESTED** [review]

File: src/modules/hyprland/workspaces.cpp
Comment (zjeffer) [inline src/modules/hyprland/workspaces.cpp:180]: @MightyPlaza why was this substr(13) and not 8?

File: src/modules/hyprland/workspaces.cpp
Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp:321]: Do you mean just remove the `fmt::arg("id", id())` from this line?

Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp]: yes

File: src/modules/hyprland/workspaces.cpp
Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp:180]: yes it should (maybe I was thinking "special:name:<name_here>" but that isn't what happens)

[special.txt](https://github.com/Alexays/Waybar/files/12177873/special.txt)
this diff fixes it and issues I found in the the on_click function (no idea how I didn't notice this before)

File: src/modules/hyprland/workspaces.cpp
Comment (MightyPlaza) [inline src/modules/hyprland/workspaces.cpp:227]: only give id as 0 if name is not int, otherwise use name as id

also workspace_name == "special" since named special workspaces have the same issue as the normal named workspaces

File: src/modules/hyprland/workspaces.cpp
Comment (MightyPlaza) [inline src/modules/hyprland/workspaces.cpp:241]: why use id here instead of desc?

File: src/modules/hyprland/workspaces.cpp
Comment (MightyPlaza) [inline src/modules/hyprland/workspaces.cpp:334]: what the order we are trying to go for here?

File: src/modules/hyprland/workspaces.cpp
Comment (MightyPlaza) [inline src/modules/hyprland/workspaces.cpp:384]: persistent workspaces will also have id > 0

**Review by MightyPlaza — CHANGES_REQUESTED** [review]

File: src/modules/hyprland/workspaces.cpp
Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp:241]: Because I use this ID to calculate the IDs the persistent workspaces should get.

I [forked](https://github.com/zjeffer/split-monitor-workspaces) Duckonaut's `split-monitor-workspaces` Hyprland plugin to make it possible to automatically set up keyboard shortcuts to switch to the xth workspace on the current focused monitor. The plugin calculates the IDs the workspaces should get based on the monitor ID, and here I used the same formula (because the IDs have to match).

Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp:241]: would make more sense to use ```workspace name:<monitor-desc-here>_<workspace-number-here>```, monitor id can change on restart but monitor desc is persistent.
Also it's more consistent with the rest of hyprland

File: src/modules/hyprland/workspaces.cpp
Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp]: Ok, I implemented the first option.

File: src/modules/hyprland/workspaces.cpp
Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp:252]: But how can I then ensure both waybar and the plugin give the same name to the workspaces?

Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp:252]: waybar should only use the name you set in the config file

File: src/modules/hyprland/workspaces.cpp
Comment (MightyPlaza) [inline src/modules/hyprland/workspaces.cpp:357]: id of named special are from -98 to -1

**Review by MightyPlaza — APPROVED** [review]
LGTM

File: src/modules/hyprland/workspaces.cpp
Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp:252]: What if you want waybar to work with any monitor you connect, without having to specify the name in the config file?

For example, if I connect my laptop to a monitor it has never seen before, waybar and my fork of the plugin will work together to automatically show persistent workspaces and set up the correct keybinds.

If we did it with monitor descriptions, you would have to do all that stuff manually.

Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp:252]: use the port name then

Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp:252]: I don't understand why it would be any different with the port name instead of the monitor description. I would still have to manually specify the port name in the configs.

In [the plugin](https://github.com/zjeffer/split-monitor-workspaces/blob/main/src/main.cpp#L63), I use the formula `workspace_name = (monitor_ID * amount_of_workspaces) + index + 1`. Here, I use the same formula so that any monitor I connect, both the plugin and the monitor will assign the same index to the same workspace ID (with 5 workspaces per monitor, the 4th workspace on the monitor with ID 2 will have workspace ID = 2*5 + 3 + 1 = 15).

I don't see how I can implement the same functionality with port names or monitor desc. I'm open to suggestions.

Comment (MightyPlaza) [reply src/modules/hyprland/workspaces.cpp:252]: simply use named workspaces, containing the monitor name and a number (```name:<monitor-desc-here>_<workspace-number-here>```)
this way you wouldn't rely on using initial id, which depends on the order you connect the monitors

if you want every monitor you connect on DP-1 (for example) to have the same configuration you would use DP-1 in the config
if you want each monitor you connect on DP-1 (for example) to have a different config you could use the monitor description

using id gives you much less control than either of these options

Comment (zjeffer) [reply src/modules/hyprland/workspaces.cpp:252]: I'm starting to understand. But I think changing it to use monitor descriptions might be more appropriate for another PR, as a lot of things would need to be refactored for it to work. Not to mention the changes I would have to make to the plugin.

I think we can merge this if you agree? CC @Alexays

**Review by Alexays — APPROVED** [review]

File: src/modules/hyprland/workspaces.cpp
Comment (aaronshahriari) [reply src/modules/hyprland/workspaces.cpp:252]: Did anyone create a PR for using the monitor description? My monitor IDs change quite frequently and I would love to be able to set specific monitor descriptions to the output for my waybar as I do in my hyprland configuration.

Thanks!

**Pull request merged**
